sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java

```diff
@@ -424,13 +465,40 @@ public void processElement(
         BoundedWindow window,
         OutputReceiver<KV<K, Iterable<InputT>>> receiver) {
       LOG.debug("*** BATCH *** Add element for window {} ", window);
+      if (shouldCareAboutWeight()) {
+        final long elementWeight = weigher.apply(element.getValue());
+        if (elementWeight + storedBatchSizeBytes.read() > batchSizeBytes) {
```

Comment:
this defeats the readLater optimization, since you're eagerly reading the value here (meaning also there's no point in the below readLater). you should add readLaters for minBufferedTs (if needed) and storedBatchSize earlier in the function.